JBTM

tomjenkinson opened the pull request:
@mswiderski please do let me know the protocol to get this merged.

mswiderski commented:
ok to test

mswiderski commented:
@tomjenkinson thanks for the PR.

in general it needs to be reviewed and build green, that's all :)

MarianMacik commented on jbpm-case-mgmt/jbpm-case-mgmt-impl/src/test/java/org/jbpm/casemgmt/impl/util/PoolingDataSource.java:
```diff
@@ -0,0 +1,148 @@
+//            Connection connection = xaConnection.getConnection();
+//            connection.prepareStatement("DROP ALL OBJECTS");
+//            connection.commit();
+//            xaConnection.close();
```
Please remove commented code if it's not needed. Thanks!

tomjenkinson replied:
Thanks - I have removed this now

MarianMacik commented on jbpm-human-task/jbpm-human-task-core/src/test/java/org/jbpm/services/task/DeadlinesBaseTest.java:
```diff
@@ -409,24 +409,24 @@ public void testDelayedEmailNotificationOnDeadlineTaskCompletedMultipleTasks() t
+
           taskService.addTask(task, new HashMap<String, Object>());
-          taskService.addTask(task2, new HashMap<String, Object>());
+          taskService.addTask(task2, new HashMap<String, Object>());  // TODO - the TSR is accessed by drools during afterCompletion
```
Will this be done in another PR?

tomjenkinson replied:
Sorry, I can remove that - it was addressed in my changes to the  Synchronization. I will remove the wording.

tomjenkinson commented:
I agree with CloseSafeMemoryContextFactory. I can try to refactor those a bit but as they are in the test package it can be tricky to share them but I think I maybe just find a different JNDI provider - that is all I needed it for.

Regarding PDS I can try to refactor that too...

mswiderski commented:
jenkins retest this

mswiderski commented:
Jenkins retest this

mswiderski commented:
looks like the failures are related to remove of PoolingDataSource/bitronix in jbpm-test module

manstis commented:
What about all the REST failing tests on [kie-wb-distributions](https://kie-jenkins.rhev-ci-vms.eng.rdu2.redhat.com/job/kie-wb-distributions/) (since 4th July) can those failures also possibly be related!?

mswiderski commented:
@manstis this change is only to replace bitronix in jbpm with narayana so it has no impact on workbench at all. Moreover in v7 we don't have any persistence need in workbench either thus no need for anything related to this PR

tomjenkinson commented:
So, I see some failures but I am not too sure about them.

There is still something I think I need to do in ./jbpm-container-test/jbpm-in-container-test/jbpm-container-test-suite as it still refers to BTM in there but the failure on your CI seems to be different. I will try to run these tests locally - I haven't ran them yet as I was just doing "mvn clean install" - I guess I needs some flags for the integration tests.

tsurdilo commented:
@tomjenkinson to run the integration tests you do with
mvn clean install -Pwildfly10

to run a specific integration test you can do

mvn clean install -Pwildfly10 -Dit.test=myTestName

thanks

tomjenkinson commented:
Thanks! I will try that now.

tomjenkinson commented:
For some reason I can't access the CI link today? It seems to need credentials now and I don't think I needed them to access it previously.

manstis commented:
@tomjenkinson The authentication issue has been reported to support by @mbiarnes

MarianMacik commented:
@tomjenkinson @mswiderski please see #902 for explanation regarding jbpm-remote-ejb-test issues.

Regarding the jbpm-in-container-test module, BTM is only used with Tomcat since it does not support JTA out of the box. Therefore I would expect changes only in Tomcat profile. Is there any reason the datasource name was changed with every container?

Thanks

MarianMacik commented on jbpm-container-test/jbpm-in-container-test/jbpm-container-test-suite/pom.xml:
```diff
@@ -101,6 +96,12 @@
+      <!--<artifactId>hibernate-core</artifactId>-->
+      <!--<scope>provided</scope>-->
+      <!--<scope>provided</scope>-->
+    <!--</dependency>-->
```
Why was this dependency commented?

tomjenkinson replied:
oversight - I will revert

tomjenkinson commented:
I will revert the non-tomcat profile changes

MarianMacik commented on jbpm-container-test/jbpm-in-container-test/jbpm-container-test-suite/src/test/resources/org/jbpm/test/container/archive/localtransactions/tomcat-context.xml:
```diff
@@ -6,10 +6,16 @@
+              factory="org.h2.jdbcx.JdbcDataSourceFactory"/>
+    <Resource name="jdbc/jbpm" uniqueName="transactionalDataSource" auth="Container" type="javax.sql.DataSource"
+              username="sa" password="sa" driverClassName="com.arjuna.ats.jdbc.TransactionalDriver"
+              url="jdbc:arjuna:java:comp/env/h2DataSource" description="Transactional Driver Data Source"/>
```
Why are there 2 datasources?

I suspect this will only work with H2 since I don't see any filtering as previously where there were org.jbpm.datasource.* Maven properties. We need to test with any DBs, not only with H2.

tomjenkinson replied:
The com.arjuna.ats.jdbc.TransactionalDriver datasource wraps the h2DataSource - it should be able to wrap any DS type.

MarianMacik replied:
OK so can this be made a filtered resource so we can use e.g. PostgreSQL with it like before?

tomjenkinson replied:
I can certainly try, I am not sure how the filtered resources work right now

MarianMacik replied:
Please look on two files above btm-resources.properties and btm-config.properties, they were filtered resources.

tomjenkinson replied:
I have looked at this and it is still not clear to me how the filtered-resources work sorry :(

I tracked references of the file btm-resources.properties to bitronix.resources.properties in the pom.xml, I can see that is referenced then in btm-config.properties via the property bitronix.tm.resource.configuration. btm-config is referenced back in the pom by the properties bitronix.config.properties which sets a system property bitronix.tm.configuration.

I therefore assume that BTM itself is processing those files so I just not sure what is needed here.

Let me explain the config I have put in the tomcat-context.xml and then you can let me know what you think is best.

Narayana wraps XA datasources within the com.arjuna.ats.jdbc.TransactionalDriver. So in the config I have here I need a standard H2 datasource (h2DataSource), then I need to reference that wrapped by the com.arjuna.ats.jdbc.TransactionalDriver. Arjuna registers for jdbc:arjuna URLS and the rest of the URL is then used to determine the URL of the datasource we need to wrap.

Is this tomcat-context.xml filtered at all I wonder? Perhaps I can use ${org.jbpm.datasource.connection.url} for the H2 URL and ${org.jbpm.datasource.username} and so on directly in that XML file and they might be replaced at build time?

tomjenkinson replied:
I think I can put those properties in this file and they seem to be filtered

tomjenkinson replied:
Yeah. That worked - thanks for the feedback.

MarianMacik replied:
Sorry, for the late response. Yes, resources are filtered in pom.xml file. Thanks!

mswiderski commented:
@MarianMacik we need to do similar as we did for KIE Server - externalise the actual data source for tomcat via context.xml. See point 5 in this instructionhttps://raw.githubusercontent.com/kiegroup/droolsjbpm-integration/master/kie-server-parent/kie-server-wars/kie-server/src/main/webc-resources/README.md

that way we can have context.xml per db that most likely would be a filtered resource so we could add various properties in there.

MarianMacik commented:
@mswiderski well yes I think that if we move it to filtered-resources and add those maven properties to it it should work. Is this what you meant?

mswiderski commented:
@MarianMacik yes, that is what I tried to say :)

tomjenkinson commented:
I am having trouble with the container-test. For some reason it is not setting the EMF. I am looking into that.
More fundamentally though I need to provide a web.xml only for the tomcat8 profile. Is there something I can do to ensure a web.xml is provided only in the tomcat profile?

MarianMacik commented:
Well I am seeing only jbpm-remote-ejb-test module failing, but you didn't do any changes to it. Or did I miss anything?

tomjenkinson commented:
I am not sure the tomcat8 profile runs on CI? It is not working locally for me which is what I am going on but the reason it's not working is different reason to the missing web.xml (it gets slf4j conflicts or the EMF is not installed) so I can't verify the need of the web.xml yet. 

Those remote-ejb-tests should have passed IMO as I rebased on your earlier https://github.com/kiegroup/jbpm/pull/902 and as you say I made no edits there now.

tomjenkinson commented:
I checked the output on the CI box and do not see the tomcat8 profile running there so that is why CI tests for the web.xml change I need are not failing.

I have no idea why the remote EJB tests failed. I have kicked off another build (I think) by commit --amend and push -f it. Hopefully those will pass now. I definitely have #902 in my branch.

MarianMacik commented:
Yes, exactly, I am talking about those remote ejb tests which haven't passed on CI, but they should. It is strange since there were no modifications.

Regarding the tomcat and web.xml. Right now it is copied every time regardless of the used container, which is OK. It is done so [here](https://github.com/tomjenkinson/jbpm/blob/91c0aab3801943198638e3dd36416c2f672661c4/jbpm-container-test/jbpm-in-container-test/jbpm-container-test-suite/src/test/java/org/jbpm/test/container/archive/LocalTransactions.java#L90-L90).

BTW tests are not run using -P parameter. There is a [readme](https://github.com/tomjenkinson/jbpm/tree/91c0aab3801943198638e3dd36416c2f672661c4/jbpm-container-test/jbpm-in-container-test) for these tests.
The `container.profile` property is here because it activates shrinkwrap dependencies in `shrinkwrap-war-profiles`. Just run `mvn clean install -Dcontainer.profile=tomcat8` from `jbpm-in-container-test` directory which acts as a root for all submodules.

tomjenkinson commented:
I don't think I can add the same web.xml to all containers. The web.xml will be something like this:
https://github.com/jbosstm/quickstart/blob/master/transactionaldriver-and-tomcat/src/main/webapp/WEB-INF/web.xml

I will only be adding the jar that has org.jboss.narayana.tomcat.jta.NarayanaJtaServletContextListener in it to the tomcat profile so on other containers (e.g. wildfly10) I will get a class not found issue (at least that is what I had when I had the web.xml earlier in my testing)

MarianMacik commented:
Sorry, I was thinking about tomcat context.xml file. Well I guess you will have to do it in the code by checking port probably. But when I run `mvn clean install -Dcontainer.profile=tomcat8` I am getting
``` Caused by: java.lang.ClassNotFoundException: bitronix.tm.resource.ResourceObjectFactory ```
Which is there because it is configured in https://github.com/tomjenkinson/jbpm/blob/master/jbpm-container-test/jbpm-in-container-test/jbpm-container-test-suite/pom.xml#L698 so what is the equivalent for Narayana/JBTM? I think we should start here.

tomjenkinson commented:
I have a code around for the servley context now (I hope): https://github.com/tomjenkinson/jbpm/commit/95c20f69661e50d2229fc857451617c0bdec179f

I will try mvn clean install -Dcontainer.profile=tomcat8 here to see what I can do and then look at the equivalent for bitronix.tm.resource.ResourceObjectFactory

tomjenkinson commented:
What file do you see Caused by: java.lang.ClassNotFoundException: bitronix.tm.resource.ResourceObjectFactory in? The war is failing to deploy for me but jbpm-container-test/jbpm-in-container-test/jbpm-container-test-suite/log.dir_IS_UNDEFINED/server.log is empty :(

mswiderski commented:
@tomjenkinson I think you can avoid to modify web.xml if you annotate the listener class with http://docs.oracle.com/javaee/7/api/javax/servlet/annotation/WebListener.html that way it will be directly loaded by tomcat once it's on class path

tomjenkinson commented:
OK, I take a look at that. I think that https://github.com/bitronix/btm/blob/master/btm/src/main/java/bitronix/tm/resource/ResourceObjectFactory.java is kind of how btm does the wrapping of the driver which we do in the tomcat-context.xml so maybe that line can be removed from the cargo config?

MarianMacik commented:
@tomjenkinson I see it on the command line. [Here](https://pastebin.com/wrt3N91v) it is.

mswiderski commented:
maybe cargo config we use for kie server can help here 
https://github.com/kiegroup/droolsjbpm-integration/blob/master/kie-server-parent/kie-server-tests/pom.xml#L367
https://github.com/kiegroup/droolsjbpm-integration/blob/master/kie-server-parent/kie-server-tests/kie-server-integ-tests-jbpm/pom.xml#L321

tomjenkinson commented:
At the moment I can't test it as my app fails to deploy but I think just removing that line from the cargo config related to objectfactory should work as the resource is defined in the tomcat-context.xml

tomjenkinson commented:
I got it to deploy now (something about that workaround I had) - tests are passing here. However the wildfly10 won't work as I have that web.xml for all configs now. I will try with http://docs.oracle.com/javaee/7/api/javax/servlet/annotation/WebListener.html to see if I can get it to work without web.xml.

MarianMacik commented:
@mswiderski the config you posted is for H2 only since kie-server is not prepared for production DBs in community. We have separate configuration which we use for our tests, but jbpm-in-container tests config is more or less the same as our internal one. So we will see. At least we have another source of configuration.

Anyway, I would prefer the profiles way which you posted in your config here:
https://github.com/kiegroup/droolsjbpm-integration/blob/master/kie-server-parent/kie-server-tests/kie-server-integ-tests-jbpm/pom.xml#L321

And why there is no such listener in kie-server tests?

tomjenkinson commented:
I managed to get the web.xml change removed as you say with a @WebListeiner so that seems fine now. I did a mvn deploy of Narayana snapshot with that working so lets see what happens...

MarianMacik commented:
Great! Btw ignore my previous post since it was with context.xml file, but we are now talking about web.xml :D

MarianMacik commented:
@tomjenkinson I have found that you deleted datasource from cargo plugin - so I don't think it will work now.

tomjenkinson commented:
OK, I will take a look. It is failing on the Wildfly version - likely the cargo config is required for other containers but because of my tomcat-context.xml it is not needed on Tomcat.

tomjenkinson commented:
I hadn't pushed the change to remove the web.xml - I had only checked it locally.

On my machine wildfly and tomcat8 are both working with the cargo set up as seen. Hopefully the current CI build will verify this....

The cargo section I removed only affects tomcat and this datasource is now defined in the tomcat-context.xml.

tomjenkinson commented:
Great - it seems to all be working except those remote ejb failures. I will try to get to the bottom of those now.

tomjenkinson commented:
They might be something to do with serialization: https://developer.jboss.org/thread/199036

tomjenkinson commented:
I think there is no session in the EMF but weirdly when I run without my changes the test passes but it doesn't even look to deploy from the log :(

tomjenkinson commented:
I noticed a difference. When the test is working the ejb-services-app.war/WEB-INF/lib/jbpm-kie-services-7.1.0-SNAPSHOT.jar is an empty folder not a Jar - which is strange!

tomjenkinson commented:
Ignore that sorry, I was looking at a partial build failure when I saw kie-services was empty

tomjenkinson commented:
I found that I need to enable lazy load sessions: https://github.com/kiegroup/jbpm/pull/898/commits/f6107a3e73bed9bc18d6dd2b641dd75207b318c9

It does seem reasonable to need that as the session is not accessed before it is returned. I wonder if somehow BTM was touching the object to initialize the collection and that was the reason why it was needed before? Though BTM would not have been used as a transaction manager in WFLY so it is strange. I guess I rather say I don't understand why it was ever not needed...

tomjenkinson commented:
With my latest changes this all works in my CI. For some reason the kiegroup box can't pull my changes down? @mswiderski

mswiderski commented:
@tomjenkinson will check it out tomorrow why it does not find it

MarianMacik commented:
@tomjenkinson I have cloned your repo and tried to run ejb remote test _without_ your change to ```persistence.xml``` using ```mvn clean install -Dcontainer.profile=wildfly10``` and it all passed. I really cannot see any connection between `java.io.EOFException: Read past end of file` and this Hibernate property. Do we really need to turn on that property since it is a bad practice? Thanks.

tomjenkinson commented:
I agree BTM did not (seem to) need that configuration option - I have observed the same here in my own CI system.

The "java.io.EOFException: Read past end of file" is raised because the TaskImpl that is trying to be read by jboss remoting fails and the Exception that JBoss remoting then raises is not serializable so that is coming from JBR effectively.

JBoss remoting is getting an exception because TaskImpl is externalizable and trying to write out the delegates which have not been loaded yet within the em::find(). The externalizable call needs to call bag.size() which HHH then tries to load but at the point JBR is processing this class there is no HHH session. 

In other words, the lazy load setting is a requirement of the TaskImpl implementation because the collection is not fully loaded before the transaction completes and the result is returned when there is no HHH session (it seems).

So to me this makes sense to need the setting, I am not sure why you don't need the setting before. If you have the test in the debugger (without my changes) perhaps you can breakpoint https://github.com/kiegroup/jbpm/blob/master/jbpm-human-task/jbpm-human-task-jpa/src/main/java/org/jbpm/services/task/impl/model/TaskImpl.java#L175 and see when that is called. For me it is called when there is no HHH session (i.e. after the TX)

tomjenkinson commented:
@mswiderski - it seems to be finding the repo OK now thanks

MarianMacik commented:
@tomjenkinson But jbpm-remote-ejb is not run with bitronix or narayana on classpath. It is run only with WildFly and it is passing for me **_without_** the change in persistence.xml. Can you try that too?

tomjenkinson commented:
sure, I will try to work out what the difference is. As I say, on my CI I see that the unpatched version does not have that issue. But what I can also say is I understand why lazy load should be necessary if that collection is not touched before it is returned.

tomjenkinson commented:
I think there is a compilation error in JBPM at the moment due to something like a drools upgrade it seems?

[ERROR] <http://narayanaci1.eng.hst.ams2.redhat.com/job/jbpm/ws/jbpm-persistence/jbpm-persistence-jpa/src/test/java/org/jbpm/persistence/session/VariablePersistenceStrategyTest.java>:[559,56] non-static method getClassIdValue(java.lang.Object) cannot be referenced from a static context

mswiderski commented:
@tomjenkinson looking at the compilation issue at the moment

MarianMacik commented on jbpm-runtime-manager/src/test/resources/kjar/jbpm-module (1).iml:
```diff
@@ -0,0 +1,12 @@
+    <orderEntry type="inheritedJdk" />
+    <orderEntry type="sourceFolder" forTests="false" />
+  </component>
+</module>
```
Also, please remove this file from commit since it is only for IDE.

mswiderski commented:
fix for compilation error is here https://github.com/kiegroup/jbpm/pull/908

MarianMacik commented on jbpm-services/jbpm-services-cdi/src/test/resources/kjar-jpa/persistence-test (2).iml:
```diff
@@ -0,0 +1,12 @@
+    <orderEntry type="inheritedJdk" />
+    <orderEntry type="sourceFolder" forTests="false" />
+  </component>
+</module>
```
Same as before - IDE file.

MarianMacik commented on jbpm-services/jbpm-services-ejb/jbpm-services-ejb-client/src/test/resources/kjar/custom-data-project.iml:
```diff
@@ -0,0 +1,12 @@
+    <orderEntry type="inheritedJdk" />
+    <orderEntry type="sourceFolder" forTests="false" />
+  </component>
+</module>
```
Same as before - IDE file.

MarianMacik commented:
@tomjenkinson OK, I think I may have found the root cause: There is a change in https://github.com/kiegroup/jbpm/pull/898/files#diff-8acfbfff2711aaf51b103a385a64750e  -> TaskTransactionInterceptor. Can you please revert those changes back and run it on our KIE Jenkins so we will see if remote-ejb tests will still fail or will pass? I don't know why changes were made there but it may have changed the way the engine behaves now since this.service.tpm.endCommandScopedEntityManager(); is not called anymore in afterCompletion() method. @mswiderski wdyt? Thanks.

tomjenkinson commented:
@MarianMacik the reason for those changes is because you can't access the TSR in aftercompletion. Arjuna enforces this strictly so endCommandScopedEntityManager would fail

mswiderski commented:
@tomjenkinson  if that is the case why it does work on WildFly/EAP at the moment?

mswiderski commented:
moreover when looking at the code it should also fail for other parts as exactly same code is used for processes - org.drools.persistence.jpa.AbstractPersistenceContextManager.endCommandScopedEntityManager() is the method that is actually called here. but then a process without user tasks is also calling this method https://github.com/kiegroup/drools/blob/master/drools-persistence/drools-persistence-jpa/src/main/java/org/drools/persistence/PersistableRunner.java#L494

MarianMacik commented:
@mswiderski @tomjenkinson I have confirmed it by running jbpm-remote-ejb-test locally. Once with master version of jbpm-human-task-jpa - tests have passed and once with jbpm-human-task-jpa from this PR - tests have failed. So it is not Narayana/BTM change, rather it is TaskTransactionInterceptor change which causes jbpm-remote-ejb tests to fail.

BTW I have found this change too: https://github.com/kiegroup/jbpm/pull/898/files#diff-268b3f2ef2189c820758c35ba5983f82 Not sure if it is a fix indeed - what would happen if the scheduler can't be shutted down for whatever reason? Won't it permanently stuck in a loop? Or how long do we want to wait?

tomjenkinson commented:
I have split out the changes that are functional issues not strictly related to the Narayana upgrade into separate commits.

@mswiderski at the moment the closeEntityManager is _not_ happening in WFLY (it might be why the test passes actually without my changes). the tpm.endCommandScopedEntityManager() is a no-op in WFLY atm because the TSR is returns null for your key so it does not clear the EM.

mswiderski commented:
@tomjenkinson correct, the entity manager is not cleared in that case but this line is invoked https://github.com/kiegroup/drools/blob/master/drools-persistence/drools-persistence-jpa/src/main/java/org/drools/persistence/jpa/AbstractPersistenceContextManager.java#L120

Since that code already handles case when there is no transaction so TSR will not be available (to be more precise its resources) then it returns null but will still clean up the ksession environment. So I would say we should keep the call to the endCommandScopedEntityManager in there as it should not cause any problems. Did you get any errors when running this without your modification of this commit : https://github.com/kiegroup/jbpm/pull/898/commits/981393739b82d982f07b5b3ef799608a5f126178

tomjenkinson commented:
Yes, I did get a problem. One of the tests said that a collection is still associated with a previous session.

DeadlinesLocalTest>DeadlinesBaseTest.testDelayedEmailNotificationOnDeadlineTaskCompletedMultipleTasks:429

```
javax.persistence.PersistenceException: org.hibernate.HibernateException: Illegal attempt to associate a collection with two open sessions. Collection : [org.jbpm.services.task.impl.model.TaskImpl.peopleAssignments.businessAdministrators#1]
	at org.hibernate.jpa.spi.AbstractEntityManagerImpl.convert(AbstractEntityManagerImpl.java:1692) ~[hibernate-entitymanager-5.1.4.Final.jar:5.1.4.Final]
	at org.hibernate.jpa.spi.AbstractEntityManagerImpl.convert(AbstractEntityManagerImpl.java:1602) ~[hibernate-entitymanager-5.1.4.Final.jar:5.1.4.Final]
	at org.hibernate.jpa.spi.AbstractEntityManagerImpl.convert(AbstractEntityManagerImpl.java:1608) ~[hibernate-entitymanager-5.1.4.Final.jar:5.1.4.Final]
	at org.hibernate.jpa.spi.AbstractEntityManagerImpl.persist(AbstractEntityManagerImpl.java:1152) ~[hibernate-entitymanager-5.1.4.Final.jar:5.1.4.Final]
	at org.jbpm.services.task.persistence.JPATaskPersistenceContext.persistTask(JPATaskPersistenceContext.java:125) ~[classes/:na]
	at org.jbpm.services.task.impl.TaskInstanceServiceImpl.addTask(TaskInstanceServiceImpl.java:109) ~[classes/:na]
	at org.jbpm.services.task.commands.AddTaskCommand.execute(AddTaskCommand.java:117) ~[classes/:na]
	at org.jbpm.services.task.commands.AddTaskCommand.execute(AddTaskCommand.java:54) ~[classes/:na]
	at org.jbpm.services.task.commands.TaskCommandExecutorImpl$SelfExecutionCommandService.execute(TaskCommandExecutorImpl.java:76) ~[classes/:na]
	at org.jbpm.services.task.commands.TaskCommandExecutorImpl$SelfExecutionCommandService.execute(TaskCommandExecutorImpl.java:63) ~[classes/:na]
	at org.drools.core.command.impl.AbstractInterceptor.executeNext(AbstractInterceptor.java:39) [drools-core-7.1.0-20170711.111738-223.jar:7.1.0-SNAPSHOT]
	at org.jbpm.services.task.persistence.TaskTransactionInterceptor.execute(TaskTransactionInterceptor.java:69) [classes/:na]
	at org.jbpm.services.task.persistence.TaskTransactionInterceptor.execute(TaskTransactionInterceptor.java:45) [classes/:na]
	at org.drools.core.command.impl.AbstractInterceptor.executeNext(AbstractInterceptor.java:39) [drools-core-7.1.0-20170711.111738-223.jar:7.1.0-SNAPSHOT]
	at org.drools.persistence.jta.TransactionLockInterceptor.execute(TransactionLockInterceptor.java:73) [drools-persistence-api-7.1.0-20170711.112256-223.jar:7.1.0-SNAPSHOT]
	at org.drools.persistence.jta.TransactionLockInterceptor.execute(TransactionLockInterceptor.java:45) [drools-persistence-api-7.1.0-20170711.112256-223.jar:7.1.0-SNAPSHOT]
	at org.drools.core.command.impl.AbstractInterceptor.executeNext(AbstractInterceptor.java:39) [drools-core-7.1.0-20170711.111738-223.jar:7.1.0-SNAPSHOT]
	at org.drools.persistence.jpa.OptimisticLockRetryInterceptor.internalExecute(OptimisticLockRetryInterceptor.java:102) [drools-persistence-jpa-7.1.0-20170711.112731-222.jar:7.1.0-SNAPSHOT]
	at org.drools.persistence.jpa.OptimisticLockRetryInterceptor.execute(OptimisticLockRetryInterceptor.java:83) [drools-persistence-jpa-7.1.0-20170711.112731-222.jar:7.1.0-SNAPSHOT]
	at org.drools.persistence.jpa.OptimisticLockRetryInterceptor.execute(OptimisticLockRetryInterceptor.java:44) [drools-persistence-jpa-7.1.0-20170711.112731-222.jar:7.1.0-SNAPSHOT]
	at org.drools.core.runtime.InternalLocalRunner.execute(InternalLocalRunner.java:37) [drools-core-7.1.0-20170711.111738-223.jar:7.1.0-SNAPSHOT]
	at org.drools.core.runtime.InternalLocalRunner.execute(InternalLocalRunner.java:41) [drools-core-7.1.0-20170711.111738-223.jar:7.1.0-SNAPSHOT]
	at org.jbpm.services.task.commands.TaskCommandExecutorImpl.execute(TaskCommandExecutorImpl.java:47) [classes/:na]
	at org.jbpm.services.task.impl.command.CommandBasedTaskService.addTask(CommandBasedTaskService.java:298) [classes/:na]
	at org.jbpm.services.task.DeadlinesBaseTest.testDelayedEmailNotificationOnDeadlineTaskCompletedMultipleTasks(DeadlinesBaseTest.java:429) [test-classes/:na]
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method) ~[na:1.8.0_121]
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62) ~[na:1.8.0_121]
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43) ~[na:1.8.0_121]
	at java.lang.reflect.Method.invoke(Method.java:498) ~[na:1.8.0_121]
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:50) [junit-4.12.jar:4.12]
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12) [junit-4.12.jar:4.12]
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:47) [junit-4.12.jar:4.12]
	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17) [junit-4.12.jar:4.12]
	at org.junit.internal.runners.statements.RunBefores.evaluate(RunBefores.java:26) [junit-4.12.jar:4.12]
	at org.junit.internal.runners.statements.RunAfters.evaluate(RunAfters.java:27) [junit-4.12.jar:4.12]
	at org.junit.runners.ParentRunner.runLeaf(ParentRunner.java:325) [junit-4.12.jar:4.12]
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:78) [junit-4.12.jar:4.12]
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:57) [junit-4.12.jar:4.12]
	at org.junit.runners.ParentRunner$3.run(ParentRunner.java:290) [junit-4.12.jar:4.12]
	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:71) [junit-4.12.jar:4.12]
	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:288) [junit-4.12.jar:4.12]
	at org.junit.runners.ParentRunner.access$000(ParentRunner.java:58) [junit-4.12.jar:4.12]
	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:268) [junit-4.12.jar:4.12]
	at org.junit.internal.runners.statements.RunBefores.evaluate(RunBefores.java:26) [junit-4.12.jar:4.12]
	at org.junit.internal.runners.statements.RunAfters.evaluate(RunAfters.java:27) [junit-4.12.jar:4.12]
	at org.junit.runners.ParentRunner.run(ParentRunner.java:363) [junit-4.12.jar:4.12]
	at org.junit.runner.JUnitCore.run(JUnitCore.java:137) [junit-4.12.jar:4.12]
	at com.intellij.junit4.JUnit4IdeaTestRunner.startRunnerWithArgs(JUnit4IdeaTestRunner.java:68) [junit-rt.jar:na]
	at com.intellij.rt.execution.junit.IdeaTestRunner$Repeater.startRunnerWithArgs(IdeaTestRunner.java:51) [junit-rt.jar:na]
	at com.intellij.rt.execution.junit.JUnitStarter.prepareStreamsAndStart(JUnitStarter.java:242) [junit-rt.jar:na]
	at com.intellij.rt.execution.junit.JUnitStarter.main(JUnitStarter.java:70) [junit-rt.jar:na]
Caused by: org.hibernate.HibernateException: Illegal attempt to associate a collection with two open sessions. Collection : [org.jbpm.services.task.impl.model.TaskImpl.peopleAssignments.businessAdministrators#1]
	at org.hibernate.collection.internal.AbstractPersistentCollection.setCurrentSession(AbstractPersistentCollection.java:641) ~[hibernate-core-5.1.4.Final.jar:5.1.4.Final]
	at org.hibernate.event.internal.WrapVisitor.processCollection(WrapVisitor.java:49) ~[hibernate-core-5.1.4.Final.jar:5.1.4.Final]
	at org.hibernate.event.internal.AbstractVisitor.processValue(AbstractVisitor.java:104) ~[hibernate-core-5.1.4.Final.jar:5.1.4.Final]
	at org.hibernate.event.internal.WrapVisitor.processComponent(WrapVisitor.java:122) ~[hibernate-core-5.1.4.Final.jar:5.1.4.Final]
	at org.hibernate.event.internal.AbstractVisitor.processValue(AbstractVisitor.java:110) ~[hibernate-core-5.1.4.Final.jar:5.1.4.Final]
	at org.hibernate.event.internal.WrapVisitor.processValue(WrapVisitor.java:108) ~[hibernate-core-5.1.4.Final.jar:5.1.4.Final]
	at org.hibernate.event.internal.AbstractVisitor.processEntityPropertyValues(AbstractVisitor.java:59) ~[hibernate-core-5.1.4.Final.jar:5.1.4.Final]
	at org.hibernate.event.internal.AbstractSaveEventListener.visitCollectionsBeforeSave(AbstractSaveEventListener.java:354) ~[hibernate-core-5.1.4.Final.jar:5.1.4.Final]
	at org.hibernate.event.internal.AbstractSaveEventListener.performSaveOrReplicate(AbstractSaveEventListener.java:260) ~[hibernate-core-5.1.4.Final.jar:5.1.4.Final]
	at org.hibernate.event.internal.AbstractSaveEventListener.performSave(AbstractSaveEventListener.java:182) ~[hibernate-core-5.1.4.Final.jar:5.1.4.Final]
	at org.hibernate.event.internal.AbstractSaveEventListener.saveWithGeneratedId(AbstractSaveEventListener.java:113) ~[hibernate-core-5.1.4.Final.jar:5.1.4.Final]
	at org.hibernate.jpa.event.internal.core.JpaPersistEventListener.saveWithGeneratedId(JpaPersistEventListener.java:67) ~[hibernate-entitymanager-5.1.4.Final.jar:5.1.4.Final]
	at org.hibernate.event.internal.DefaultPersistEventListener.entityIsTransient(DefaultPersistEventListener.java:189) ~[hibernate-core-5.1.4.Final.jar:5.1.4.Final]
	at org.hibernate.event.internal.DefaultPersistEventListener.onPersist(DefaultPersistEventListener.java:132) ~[hibernate-core-5.1.4.Final.jar:5.1.4.Final]
	at org.hibernate.event.internal.DefaultPersistEventListener.onPersist(DefaultPersistEventListener.java:58) ~[hibernate-core-5.1.4.Final.jar:5.1.4.Final]
	at org.hibernate.internal.SessionImpl.firePersist(SessionImpl.java:778) ~[hibernate-core-5.1.4.Final.jar:5.1.4.Final]
	at org.hibernate.internal.SessionImpl.persist(SessionImpl.java:751) ~[hibernate-core-5.1.4.Final.jar:5.1.4.Final]
	at org.hibernate.internal.SessionImpl.persist(SessionImpl.java:756) ~[hibernate-core-5.1.4.Final.jar:5.1.4.Final]
	at org.hibernate.jpa.spi.AbstractEntityManagerImpl.persist(AbstractEntityManagerImpl.java:1146) ~[hibernate-entitymanager-5.1.4.Final.jar:5.1.4.Final]
	... 47 common frames omitted
```

I guess my question to you is what do you want to happen in afterCompletion? If you want an em::clear then my fix would guarantee that however as the code stands at the moment I can re-add the endCommandScopedEntityManager call as well as forcing the em::close(), I will do that initially...

This all perhaps worked for BTM as it may be more lenient on access to the TSR in AC (which is maybe why that test fails for Narayana and not BTM) and so your synchronization could retrieve the EM in AC - however BTM was wrong to allow that (or at least not portable, as AC is ran in an undefined transaction context). To me it explains why we need the lazy load in the test because the EM is cleared now which it would not have been before.

mswiderski commented:
@tomjenkinson about to test this locally, let me spend some time on the afterCompletion thingy as it might be simply an issue with test itself.

tomjenkinson commented:
Thanks - I managed to get the filtered resources working so that is good.

There is still the TODO in ExecutorImpl.java. I am not sure what you would like to do about that. It affected me during testing because the threads were not completed before the next test started and so undefined behaviour occurred but on my machine more often than not it failed

mswiderski commented:
nice we have a green build :)

@tomjenkinson I sent PR to your repo with small adjustments https://github.com/tomjenkinson/jbpm/pull/1 

so is there anything else to be tackled? Any TODOs or missing bits?

tomjenkinson commented:
I cherry-picked your commit into this branch so hopefully it is still green. It all looks OK to me :)

mswiderski commented on jbpm-audit/pom.xml:
```diff
@@ -67,6 +67,10 @@
     </dependency>
+    <dependency>
+      <groupId>org.jboss.narayana.jta</groupId>
+      <artifactId>narayana-jta</artifactId>
```
shouldn't this be in test scope? same as the removed btw was? Otherwise it will be included in all other assemblies like for EAP or WAS/Weblogic

tomjenkinson replied:
+1 - I am not sure how I missed that

MarianMacik commented on jbpm-container-test/jbpm-remote-ejb-test/jbpm-remote-ejb-test-app/src/main/resources/META-INF/persistence.xml:
```diff
@@ -68,6 +68,7 @@ http://java.sun.com/xml/ns/persistence/orm http://java.sun.com/xml/ns/persistenc
 			<!-- BZ 841786: AS7/EAP 6/Hib 4 uses new (sequence) generators which seem 
 				to cause problems -->
 			<property name="hibernate.id.new_generator_mappings" value="false" />
+			<property name="hibernate.enable_lazy_load_no_trans" value="true" />
```
Well I am still struggling with this change. :) Won't it work without this property after Maciej's changes? I am reading everywhere that this setting is not a good practice.

MarianMacik replied:
Isn't it even a bug in remote EJBs? Why do we have to "hack" this externally with the Hibernate property?

tomjenkinson replied:
It should do - you are right. I will revert it and let CI verify.

MarianMacik replied:
Thanks ;)

mswiderski commented on jbpm-human-task/jbpm-human-task-core/src/test/java/org/jbpm/services/task/DeadlinesBaseTest.java:
```diff
@@ -389,7 +389,7 @@ public void testDelayedReassignmentOnDeadline() throws Exception {
     }
 
-      @Test(timeout=10000)
+      @Test(timeout=1000000)
```
my bad as I increased that while debugging, can be go back to default 10000

tomjenkinson replied:
OK, I revert that part

mswiderski commented on jbpm-runtime-manager/src/test/java/org/jbpm/runtime/manager/concurrent/SessionTest.java:
```diff
@@ -688,7 +688,76 @@ public void run() {
+//						at com.arjuna.ats.arjuna.AtomicAction.commit(AtomicAction.java:162)
+//						at com.arjuna.ats.internal.jta.transaction.arjunacore.TransactionImple.commitAndDisassociate(TransactionImple.java:1289)
+//	... 17 more
+//						Caused by: org.hibernate.StaleStateException: Batch update returned unexpected row count from update [0]; actual row count: 0; expected: 1
```
can we maybe reduce the stack trace here and leave only the root cause info?

tomjenkinson replied:
Sure, I have trimmed this now

psiroky commented on pom.xml:
```diff
@@ -135,6 +140,21 @@
+        <groupId>org.jboss.narayana.jta</groupId>
+        <artifactId>tomcat-jta</artifactId>
+        <version>5.6.3.Final-SNAPSHOT</version>
+      </dependency>
```
We really should not be depending on 3rd party SNAPSHOTs.

(We should also move these deps into kie-parent, but that can be done later).

MarianMacik replied:
+1

tomjenkinson replied:
+1 - I can do a release of Narayana once we all agree that is all that is outstanding

tomjenkinson replied:
IMO this is the only thing we need now - once you can confirm I will do the Narayana release and roll this to non-SNAPSHOT version

mswiderski replied:
correct, I would just wait for the patch Karel (@sutaakar) was about to send for MariaDB. Once that is all in we should be good to release narayana and merge this.

@MarianMacik any other comments regarding tests on different containers/dbs or we are good there too?

MarianMacik replied:
@mswiderski See https://github.com/kiegroup/jbpm/pull/898#issuecomment-315025696

mswiderski commented on jbpm-runtime-manager/src/test/java/org/jbpm/runtime/manager/impl/factory/InMemorySessionFactoryLeakTest.java:
```diff
@@ -79,14 +83,18 @@ public InMemorySessionFactoryLeakTest(String strategy) {
+
+        new InitialContext().rebind("java:comp/UserTransaction", com.arjuna.ats.jta.UserTransaction.userTransaction());
+        new InitialContext().rebind("java:comp/TransactionManager", TransactionManager.transactionManager());
+        new InitialContext().rebind("java:comp/TransactionSynchronizationRegistry", new com.arjuna.ats.internal.jta.transaction.arjunacore.TransactionSynchronizationRegistryImple());
```
these should not be needed, I am taking a look at this test now

tomjenkinson replied:
great - it was a new one added recently and seemed to need the transaction manager available for some reason (or maybe it was TSR)

mswiderski commented:
@tomjenkinson sent new PR for some additional updates, I think these are the last :)

MarianMacik commented on jbpm-examples/pom.xml:
```diff
@@ -94,6 +94,10 @@
+      <dependency>
+          <groupId>org.jbpm</groupId>
+          <artifactId>jbpm-test-util</artifactId>
+      </dependency>
```
Is it really needed here? I don't see any Pooling Data Source in this module. But maybe it is here because of something else...

MarianMacik replied:
CloseSafeMemoryContextFactory isn't used either.

MarianMacik replied:
Ok, it is needed since it is used in org/jbpm/examples/checklist/impl/CheckListUI.java but I was unable to find it since org.jbpm.test.util.PoolingDataSource is return type of JBPMHelper.setupDataSource() and the return value is ignored so it isn't used anywhere. But maven enforcer catches this as a direct usage. So this is OK.

MarianMacik commented on jbpm-human-task/jbpm-human-task-workitems/pom.xml:
```diff
@@ -95,6 +100,11 @@
+      <groupId>org.jbpm</groupId>
+      <artifactId>jbpm-persistence-jpa</artifactId>
+      <scope>test</scope>
+    </dependency>
```
@tomjenkinson Why was this dependency added? Thanks.

tomjenkinson replied:
some are added for the transient enforcer check - I will remove it and run a build and see if it's still required

tomjenkinson replied:
seems OK without it

MarianMacik commented on jbpm-human-task/jbpm-human-task-jpa/pom.xml:
```diff
@@ -75,6 +75,11 @@
+      <groupId>org.jbpm</groupId>
+      <artifactId>jbpm-test-util</artifactId>
+      <scope>test</scope>
+    </dependency>
```
This shouldn't be needed here since there are no tests.

tomjenkinson replied:
This was picked up due to a search and replace: https://github.com/kiegroup/jbpm/tree/master/jbpm-human-task/jbpm-human-task-jpa/src/test/resources

If you are sure it's not used I will remove the src/test/resources folder and the pom updates

tomjenkinson replied:
It's here: https://github.com/kiegroup/jbpm/pull/898/commits/e1683e403aff2063ab5026e25039b0e2d4486ee7

MarianMacik replied:
Thanks, @mswiderski can you confirm it is OK?

mswiderski replied:
that seems to be correct, there are no tests at all in that module so having dependencies and configuration for transaction does not make sense.

mswiderski commented:
Jenkins execute full downstream build

tomjenkinson commented:
Great - are you OK with the commit messages and breakdown? I have merged @sutaakar PR on Narayana so I am good to go

mswiderski commented:
I would say best to squash everything into single commit and use this jira (both number and title) as commit message - https://issues.jboss.org/browse/JBPM-6171

tomjenkinson commented:
Initially I have refactored the commits to attribute per author in case attribution is important? If you prefer I can roll all authors changes into my commit.

mswiderski commented:
@tomjenkinson you've done the most of the work so all credits go to you :)

tomjenkinson commented:
Thanks - I will wait for @MarianMacik to ack your earlier question regarding any further Narayana changes required at this time.

MarianMacik commented:
OK, I think it is good to merge now. The only change with container tests was with tomcat8 and it works OK. After merge we will reconfigure our Cargo automation with @sutaakar so there might be some little changes in case we want to have 1:1 Cargo config between internal and community automation.

Just a side question - @tomjenkinson why is there jbossts-properties.xml file in some modules and in others it is not?

tomjenkinson commented:
Hmm, I can take a look at that (jbossts-properties.xml) as the narayana-jta jar does contain the same file so it is likely when I thought I needed to override something but it looks like the default one so I can likely remove it.

One thing you (/cc @mswiderski @MarianMacik @sutaakar) definitely need to be aware of is the Narayana nodeIdentifier property. Per idependent instance this value must be globally unique (and reused upon restart if you want recovery to work). You can set it in a jbossts-properties.xml file higher up the classpath or as a system property.  The standard jbossts-properties.xml does contain a value so it runs out of the box (CoreEnvironmentBean.nodeIdentifier https://github.com/jbosstm/narayana/blob/master/ArjunaJTA/narayana-jta/src/main/resources/jbossts-properties.xml#L62) but if multiple instances run without changing that property and they also share an object store (transaction log) or connect to the same XAResources then they can interfere with each other,

tomjenkinson commented:
Yeah, I remember adding those files now. I added them a year or so ago when first working on this and at that time I was using narayana-jts-idlj dependency rather than narayana-jta. Doing that meant I needed the jbossts-properties.xml for the JTA (instead of JTS) to ensure the simple config was used. I have removed them from the tests for consistency now.

mswiderski commented:
@tomjenkinson this looks ready to be merged - so I think it's time to release narayana and update the version in main pom.xml. Once that is done we'll trigger last full build and merge.

tomjenkinson commented:
Great - I have done the release, updated the version here and pushed so hopefully it still works :)

mswiderski commented:
Jenkins execute full downstream build

mswiderski commented:
Jenkins execute full downstream build

mswiderski commented:
Jenkins execute full downstream build

tomjenkinson commented:
No problem - thanks to all for the feedback!

MarianMacik commented:
Thanks, @tomjenkinson. Good job!
